File: CodeSamples/kahn.erl
-module(kahn).
-export([main/0, kahn/2]).

kahn(Graph, N) ->
    Inverse = build_inverse_graph(Graph, N),
    Stack = [V || V <- lists:seq(0, N-1), length(Inverse[V]) == 0],
    kahn(Graph, Inverse, Stack, []).

kahn(_, _, [], Order) when length(Order) == length(?GRAPH) -> lists:reverse(Order);
kahn(_, _, [], _) -> "Cyclic graph";
kahn(Graph, Inverse, [U | Stack], Order) ->
    kahn(remove_dependencies(U, Graph, Inverse), Inverse, Stack, [U | Order]).

build_inverse_graph(Graph, N) ->
    build_inverse_graph(Graph, N, dict:new()).

build_inverse_graph([], _, Inverse) -> Inverse;
build_inverse_graph([{U, Vs} | Rest], N, Inverse) ->
    NewInverse = lists:foldl(fun(V, Inv) -> dict:append(V, [U | dict:fetch(V, Inv, [])], Inv) end, Inverse, Vs),
    build_inverse_graph(Rest, N, NewInverse).

remove_dependencies(U, Graph, Inverse) ->
    {NewGraph, NewInverse} = remove_dependencies(U, Graph, Inverse, []),
    NewGraph.

remove_dependencies(_, [], Inverse, NewInverse) -> {NewInverse, NewInverse};
remove_dependencies(U, [{U, Vs} | Rest], Inverse, NewInverse) ->
    remove_dependencies(U, Rest, Inverse, NewInverse);
remove_dependencies(U, [{V, Vs} | Rest], Inverse, NewInverse) ->
    NewVs = lists:filter(fun(X) -> X /= U end, Vs),
    if
        length(NewVs) == 0 ->
            remove_dependencies(U, Rest, Inverse, dict:erase(V, NewInverse));
        true ->
            remove_dependencies(U, Rest, Inverse, dict:store(V, NewVs, NewInverse))
    end.
  
main() ->
    {ok, InputFile} = file:open("input.txt", [read]),
    {ok, OutputFile} = file:open("output.txt", [write]),
    {ok, [T | Rest]} = io:get_line(InputFile, ''),
    T1 = list_to_integer(string:strip(T, right, $\n)),
    process_cases(InputFile, OutputFile, T1, Rest),
    file:close(InputFile),
    file:close(OutputFile).

process_cases(_, _, 0, _) -> ok;
process_cases(InputFile, OutputFile, T, Rest) ->
    {ok, [NLine | Rest1]} = io:get_line(InputFile, Rest),
    N = list_to_integer(string:strip(NLine, right, $\n)),
    Graph = read_graph(InputFile, N, []),
    Order = kahn:topological_sort(Graph, N),
    case Order of
        "Cyclic graph" ->
            io:format(OutputFile, "Cyclic graph~n", []),
            process_cases(InputFile, OutputFile, T-1, Rest1);
        _ ->
            io:format(OutputFile, "~s~n", [format_order(Order)]),
            process_cases(InputFile, OutputFile, T-1, Rest1)
    end.

format_order([]) -> "";
format_order([H]) -> integer_to_list(H);
format_order([H | T]) -> integer_to_list(H) ++ " " ++ format_order(T).

read_graph(_, 0, Acc) -> lists:reverse(Acc);
read_graph(InputFile, N, Acc) ->
    {ok, [EdgeLine | Rest]} = io:get_line(InputFile, ''),
    [U, V] = string:tokens(string:strip(EdgeLine, right, $\n), " "),
    read_graph(InputFile, N - 1, [{list_to_integer(U), list_to_integer(V)} | Acc]).

File: CodeSamples/kahn.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Graph {
    int V;
    struct Node** adj;
};

struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

struct Graph* createGraph(int V) {
    struct Graph* graph = (struct Graph*)malloc(sizeof(struct Graph));
    graph->V = V;
    graph->adj = (struct Node**)malloc(V * sizeof(struct Node*));
    for (int i = 0; i < V; i++) {
        graph->adj[i] = NULL;
    }
    return graph;
}

void addEdge(struct Graph* graph, int u, int v) {
    struct Node* newNode = createNode(v);
    newNode->next = graph->adj[u];
    graph->adj[u] = newNode;
}

void topologicalSort(struct Graph* graph, int n, int* order, int* index) {
    int* inDegree = (int*)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        inDegree[i] = 0;
    }

    for (int u = 0; u < n; u++) {
        struct Node* temp = graph->adj[u];
        while (temp != NULL) {
            inDegree[temp->data]++;
            temp = temp->next;
        }
    }

    for (int u = 0; u < n; u++) {
        if (inDegree[u] == 0) {
            order[(*index)++] = u;
        }
    }

    for (int u = 0; u < n; u++) {
        struct Node* temp = graph->adj[u];
        while (temp != NULL) {
            inDegree[temp->data]--;
            if (inDegree[temp->data] == 0) {
                order[(*index)++] = temp->data;
            }
            temp = temp->next;
        }
    }
}

int* kahn(struct Graph* graph, int n) {
    int index = 0;
    int* order = (int*)malloc(n * sizeof(int));
    topologicalSort(graph, n, order, &index);

    if (index != n) {
        free(order);
        return NULL;
    }

    return order;
}

int main() {
    FILE* input_file = fopen("input.txt", "r");
    FILE* output_file = fopen("output.txt", "w");

    int t;
    fscanf(input_file, "%d", &t);

    while (t--) {
        int n;
        fscanf(input_file, "%d", &n);
        struct Graph* graph = createGraph(n);

        while (1) {
            int u, v;
            if (fscanf(input_file, "%d %d", &u, &v) == 2) {
                addEdge(graph, u, v);
            } else {
                int* result = kahn(graph, n);
                if (result == NULL) {
                    fprintf(output_file, "Cyclic graph\n");
                } else {
                    for (int i = 0; i < n; i++) {
                        fprintf(output_file, "%d", result[i]);
                        if (i < n - 1) {
                            fprintf(output_file, " ");
                        }
                    }
                    fprintf(output_file, "\n");
                }
                break;
            }
        }
    }

    fclose(input_file);
    fclose(output_file);
    return 0;
}
